### PR TITLE
fix: page re-render on form publish [OLD]

### PR DESCRIPTION
**Pull request by neatbyte-vnobis**
## Changes
Implements issue "When a form is published, re-render all the pages containing it".

## How Has This Been Tested?
Manual

## Documentation
None

### CODE REVIEW

**Comment by neatbyte-ivelychko** [comment]
Also I've implemented few more changes:
   1 - Page rerender in case that page is using the form that has been deleted. Before this change the page was trying to render form that was no longer existing, so it was crashing page. Now it shows `Form not found`;
   2 - Page rerender in case that page is using form revision that has been deleted. Same issue as above, it was crashing page. Showing `Form not found` message now;
   3 - Also updated `Delete` feature for `Form Preview` tab, now it acctually deletes entire form instead of selected revision.

**Comment by neatbyte-ivelychko** [comment]
@adrians5j this pr is ready to be reviewed.

**Comment by neatbyte-ivelychko** [comment]
@adrians5j could please check this pr

File: apps/api/graphql/src/index.ts [hunk 1]
```diff
@@ -77,6 +77,11 @@ export const handler = createHandler({
         prerenderingServicePlugins({
             eventBus: String(process.env.EVENT_BUS)
         }),
+        createFormBuilder({
```
Comment (adrians5j) [inline apps/api/graphql/src/index.ts:80]: Why this change? 🤔

File: packages/api-page-builder/src/prerendering/hooks/afterFormDeletetion.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { ContextPlugin } from "@webiny/api";
+import { PbContext } from "~/graphql/types";
+import { FormBuilderContext } from "@webiny/api-form-builder/types";
+
+export default () => {
+    return new ContextPlugin<PbContext & FormBuilderContext>(
+        async ({ pageBuilder, formBuilder }) => {
+            /**
+             * If page contains a form that has been deleted,
```
Comment (adrians5j) [inline packages/api-page-builder/src/prerendering/hooks/afterFormDeletetion.ts:9]: Looking at this code... can we have things done the other way? 

Have these event handlers in `api-form-builder`? 

This is because we wan't to keep form-related logic in its `api-form-builder` package, and not move that logic into PB package.

That way, we'll be adding PB as the dependency of FB, not FB as the dep of PB.

File: packages/app-form-builder/src/page-builder/admin/plugins/components/PeFormElement.tsx [hunk 3]
```diff
@@ -34,7 +34,8 @@ const PeForm: FormRenderer = props => {
                     } catch {
                         return apolloClient
                             .query({ query: gql(GET_PUBLISHED_FORM), variables })
-                            .then(({ data }) => data.formBuilder.getPublishedForm.data);
+                            .then(({ data }) => data.formBuilder.getPublishedForm.data)
+                            .catch(() => null);
```
Comment (adrians5j) [inline packages/app-form-builder/src/page-builder/admin/plugins/components/PeFormElement.tsx:38]: Why this change? Can we have a comment here?

File: packages/app-form-builder/src/admin/plugins/formDetails/previewContent/HeaderComponents/DeleteForm.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import * as queries from "../../../../graphql";
+import { removeFormFromListCache } from "~/admin/views/cache";
+import { usePermission } from "~/hooks/usePermission";
+import { FbRevisionModel } from "~/types";
+
+interface DeleteRevisionProps {
+    form: FbRevisionModel;
+    revision: FbRevisionModel;
+}
+const DeleteRevision: React.FC<DeleteRevisionProps> = ({ form, revision }) => {
```
Comment (adrians5j) [inline packages/app-form-builder/src/admin/plugins/formDetails/previewContent/HeaderComponents/DeleteForm.tsx:18]: I saw your comment:

>  Also updated Delete feature for Form Preview tab, now it acctually deletes entire form instead of selected revision.

What do you mean by this?

I mean, users should be able to delete just a specific revision of the form. Not sure I understand what's going on here. 🤔

File: packages/api-page-builder/src/prerendering/hooks/afterFormDeletetion.ts
Comment (adrians5j) [inline packages/api-page-builder/src/prerendering/hooks/afterFormDeletetion.ts:1]: Let's by consistent with naming. Instead of "Deletion", let's use "Delete", like we did with `afterFormRevisionDelete`.

File: packages/api-page-builder/src/prerendering/hooks/afterFormUpdate.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import { FormBuilderContext } from "@webiny/api-form-builder/types";
+
+export default () => {
+    return new ContextPlugin<PbContext & FormBuilderContext>(
+        async ({ pageBuilder, formBuilder }) => {
+            /**
+             * If page contains a form and the version of that form is "latest published revision",
+             * then it should trigger re-render for all pages that contain that form.
+             */
+            formBuilder.onFormAfterPublish.subscribe(async ({ form }) => {
```
Comment (adrians5j) [inline packages/api-page-builder/src/prerendering/hooks/afterFormUpdate.ts:12]: What about if a revision has been updated and we have a page that's showing an exact revision?

File: packages/api-page-builder/src/prerendering/hooks/afterFormUpdate.ts
Comment (adrians5j) [inline packages/api-page-builder/src/prerendering/hooks/afterFormUpdate.ts:1]: BTW the name of this file should use `Publish` verb, not `Update`.

**Review by adrians5j — CHANGES_REQUESTED** [review]
First iteration done. 

I can see one or two additional reviews happening here, but overall we're heading in the right direction.

File: packages/api-page-builder/src/prerendering/hooks/afterFormUpdate.ts (hunk 5 above)
Comment (neatbyte-ivelychko) [reply packages/api-page-builder/src/prerendering/hooks/afterFormUpdate.ts:12]: @adrians5j hm, I think that we cannot update specific form version if it has been published. Also if page shows a specific form revision and we decide to publish new revision version, the page that uses specific revision not gonna be re-rendered

File: packages/api-page-builder/src/prerendering/hooks/afterFormDeletetion.ts (hunk 2 above)
Comment (neatbyte-ivelychko) [reply packages/api-page-builder/src/prerendering/hooks/afterFormDeletetion.ts:9]: @adrians5j so you want me to move all form-builder related handlers to the api-form-builder and just import them in `api-page-builder`?

File: packages/app-form-builder/src/page-builder/admin/plugins/components/PeFormElement.tsx (hunk 3 above)
Comment (neatbyte-ivelychko) [reply packages/app-form-builder/src/page-builder/admin/plugins/components/PeFormElement.tsx:38]: If we add a form on the page and then we delete that form we get error `Form "64ee005792862d000941dc7a#0003" was not found!`. on the `Page Builder Editor` page. This is because the query in `catch` block is trying to query a FORM that does not exist. And the `catch` helps us to avoid that error.

Comment (neatbyte-ivelychko) [reply packages/app-form-builder/src/page-builder/admin/plugins/components/PeFormElement.tsx:38]: @adrians5j also It was a quick fix, I will imrove it

File: apps/api/graphql/src/index.ts (hunk 1 above)
Comment (adrians5j) [reply apps/api/graphql/src/index.ts:80]: This question still open.

File: packages/api-page-builder/src/prerendering/hooks/afterFormDeletetion.ts (hunk 2 above)
Comment (adrians5j) [reply packages/api-page-builder/src/prerendering/hooks/afterFormDeletetion.ts:9]: Let's have all of the handlers defined in the `api-form-builder` package.

Probably just moving the file into the package will be good enough (ofc you'll need to import it in relevant files).

But yeah, as mentioned, this will require us to list PB as a dep of FB, because we're accessing the `prerendering` property on the `pageBuilder` context, both of which are not defined I believe in FB's Context object.

```packages/api-form-builder/src/types.ts:438```

**Review by adrians5j — CHANGES_REQUESTED** [review]
Left some comments. Understood the UI change part (deletion of form instead of form revision). Thanks for that.

Will do another review where I analyse all of the prerendering handlers.

**Comment by neatbyte-vnobis** [comment]
Please refer to the new PR for this feature #3541
